lightning/src/routing/network_graph.rs

```diff
@@ -428,6 +428,70 @@ impl Readable for NodeInfo {
 	}
 }
 
+/// Allows for updating user's network metadata.
+pub trait UpdateMetadata {
+	/// Adds a failed route to track. If we see the route again, we'll reject it or we can take
+	/// some action (like set its fee to make route-finding not find that route again)
+	fn add_failed_route(&mut self, route: Vec<Vec<RouteHop>>);
+	/// If a route starts working again, then we can take it off the failed route list.
+	fn remove_failed_route(&mut self, route: Vec<Vec<RouteHop>>);
+	/// Returns true if a route is on the failed routes list, false otherwise.
+	fn check_route(&self, route: Vec<Vec<RouteHop>>) -> bool;
+	/// Sets the channel fee penalty being tracked in the Metadata object.
+	fn set_channel_fee_penalty(&mut self, chan_id: u64, fee_penalty: u64);
+	/// Gets a channel's fee penalty based on its channel_id (as stored in a NetworkGraph object).
+	fn get_channel_fee_penalty(&self, chan_id: u64) -> Option<&u64>;
+}
+
+/// A default metadata object that is used to implement the default functionality for the
+/// NetworkTracker trait. A user could  make their own Metadata object and extend the
```

Comment:
I think there is a question of code sharing here, if people start to work on fancy scoring logic, they may want to combine them, and having each of them implement a custom UpdateMetadata they won't be able to do so.

Reply:
I don't see why? You could write a combiner, but ultimately we *have* to have a way to linearize any kind of fancy scoring logic users want into one parameter (namely fee paid) so that we can calculate a route based on it. No matter what users do, they have to be able to linearize it.

Reply:
Yes I think we agree on the need to linearize it, my point was more on dissociating linearization method (`get_channel_fee_penalty`) from score-feeding ones (`route_succeeded`/`route_failed`). Linearization one should be on such combiner. I don't think we have to do this now, but we may have to introduce such combiner in the future.